File: baekjoon_all/03000+/boj_3154.cpp
// Solve 2025-01-18
// Update 2025-09-23

#include <bits/stdc++.h>

#define FASTIO ios_base::sync_with_stdio(false);cin.tie(NULL);
#define ALL(v) v.begin(),v.end()
#define UNIQUE(v) v.erase(unique(v.begin(),v.end()),v.end());
#define SETW(n, c) cout << setw(n) << setfill(c);
#define SETP(n) cout << fixed << setprecision(n);

using namespace std;
using ll = long long;
using uint = unsigned int;
using ull = unsigned long long;
using ld = long double;
using pii = pair<int, int>;
using pll = pair<ll, ll>;
const int INF = 0x3f3f3f3f;
const int MOD = 1000000007;

pii keyboard[10] = {
    {3, 1},
    {0, 0},
    {0, 1},
    {0, 2},
    {1, 0},
    {1, 1},
    {1, 2},
    {2, 0},
    {2, 1},
    {2, 2}
};

int get_dist(pii& p1, pii& p2) {
    return abs(p1.first - p2.first) + abs(p1.second - p2.second);
}

int main() {
    FASTIO;

    string hhmm;
    cin >> hhmm;

    int h = stoi(hhmm.substr(0, 2));
    int m = stoi(hhmm.substr(3, 2));
    int min_effort_h = h;
    int min_effort_m = m;
    int min_effort = INF;

    for (int i = h; i < 100; i += 24) {
        int d1 = i / 10;
        int d2 = i % 10;

        for (int j = m; j < 100; j += 60) {
            int d3 = j / 10;
            int d4 = j % 10;
            int effort = get_dist(keyboard[i / 10], keyboard[i % 10])
                    + get_dist(keyboard[i % 10], keyboard[j / 10])
                    + get_dist(keyboard[j / 10], keyboard[j % 10]);

            if (effort < min_effort) {
                min_effort = effort;
                min_effort_h = i;
                min_effort_m = j;
            }
        }
    }

    cout << setw(2) << setfill('0') << min_effort_h << ':';
    cout << setw(2) << setfill('0') << min_effort_m << '\n';
    return 0;
}

File: baekjoon_all/03000+/boj_3154.java
// Solve 2025-09-23

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class boj_3154 {

    static int[] x = new int[] {3, 0, 0, 0, 1, 1, 1, 2, 2, 2};
    static int[] y = new int[] {1, 0, 1, 2, 0, 1, 2, 0, 1, 2};

    static int getDist(int d1, int d2) {
        return Math.abs(x[d1] - x[d2]) + Math.abs(y[d1] - y[d2]);
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String hhmm = br.readLine();
        int h = (hhmm.charAt(0) - '0') * 10 + (hhmm.charAt(1) - '0');
        int m = (hhmm.charAt(3) - '0') * 10 + (hhmm.charAt(4) - '0');
        int hourOfMinEffort = h;
        int minuteOfMinEffort = m;
        int minEffort = 100000;

        for (int i = h; i < 100; i += 24) {
            int d1 = i / 10;
            int d2 = i % 10;
            int e = getDist(d1, d2);

            for (int j = m; j < 100; j += 60) {
                int d3 = j / 10;
                int d4 = j % 10;
                int effort = e + getDist(d2, d3) + getDist(d3, d4);

                if (effort < minEffort) {
                    minEffort = effort;
                    hourOfMinEffort = i;
                    minuteOfMinEffort = j;
                }
            }
        }

        System.out.print(hourOfMinEffort / 10);
        System.out.print(hourOfMinEffort % 10);
        System.out.print(":");
        System.out.print(minuteOfMinEffort / 10);
        System.out.println(minuteOfMinEffort % 10);
    }

}
